File: app/src/main/java/de/cineaste/android/activity/SeasonDetailActivity.java
package de.cineaste.android.activity;

import android.content.Intent;
import android.os.Bundle;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CollapsingToolbarLayout;
import android.support.v4.view.ViewPager;
import android.support.v7.app.ActionBar;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;
import android.view.View;
import android.widget.ImageView;

import com.squareup.picasso.Picasso;

import java.util.List;

import de.cineaste.android.R;
import de.cineaste.android.adapter.series.SeasonPagerAdapter;
import de.cineaste.android.database.dao.BaseDao;
import de.cineaste.android.database.dbHelper.SeriesDbHelper;
import de.cineaste.android.entity.series.Season;
import de.cineaste.android.entity.series.Series;
import de.cineaste.android.network.SeriesCallback;
import de.cineaste.android.network.SeriesLoader;
import de.cineaste.android.util.Constants;

public class SeasonDetailActivity extends AppCompatActivity {

    private Series currentSeries;
    private ImageView poster;

    private long seasonId;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_season_detail);

        SeriesDbHelper seriesDbHelper = SeriesDbHelper.getInstance(this);

        poster = findViewById(R.id.poster_image_view);

        Intent intent = getIntent();
        long seriesId = intent.getLongExtra(BaseDao.SeasonEntry.COLUMN_SEASON_SERIES_ID, -1);
        seasonId = intent.getLongExtra(BaseDao.SeasonEntry.COLUMN_SEASON_SEASON_NUMBER, -1);

        currentSeries = seriesDbHelper.getSeriesById(seriesId);

        if (currentSeries == null) {
            SeriesLoader seriesLoader = new SeriesLoader(this);
            seriesLoader.loadCompleteSeries(seriesId, new SeriesCallback() {
                @Override
                public void onFailure() {

                }

                @Override
                public void onSuccess(Series series) {
                    currentSeries = series;
                    assignData(series);
                }
            });
        } else {
            assignData(currentSeries);
        }

        initToolbar();
    }

    private void assignData(Series series) {
        SeasonPagerAdapter adapter = new SeasonPagerAdapter(getSupportFragmentManager(), series, getResources());
        ViewPager viewPager = findViewById(R.id.pager);

        viewPager.setAdapter(adapter);
        viewPager.setCurrentItem(currentSeasonIndex());
        viewPager.addOnPageChangeListener(new ViewPager.OnPageChangeListener() {
            @Override
            public void onPageScrolled(int position, float positionOffset, int positionOffsetPixels) {
                //do nothing
            }

            @Override
            public void onPageSelected(int position) {
                setPoster(position);
            }

            @Override
            public void onPageScrollStateChanged(int state) {
                //do nothing
            }
        });

        setPoster(currentSeasonIndex());
    }

    private void setPoster(int position) {
        Season season = currentSeries.getSeasons().get(position);
        final String posterPath = season.getPosterPath();

        String posterUri = Constants.POSTER_URI_SMALL
                .replace("<posterName>", posterPath)
                .replace("<API_KEY>", getString(R.string.movieKey));
        Picasso.with(this)
                .load(posterUri)
                .error(R.drawable.placeholder_poster)
                .into(poster);

        poster.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(SeasonDetailActivity.this, PosterActivity.class);
                intent.putExtra(PosterActivity.POSTER_PATH, posterPath);
                startActivity(intent);
            }
        });
    }

    private int currentSeasonIndex() {
        List<Season> seasons = currentSeries.getSeasons();
        for (int i = 0; i < seasons.size(); i++) {
            if (seasons.get(i).getId() == seasonId) {
                return i;
            }
        }

        return 0;
    }

    private void initToolbar() {
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        ActionBar actionBar = getSupportActionBar();
        if (actionBar != null)
            actionBar.setDisplayHomeAsUpEnabled(true);

        setTitleIfNeeded();
    }

    private void setTitleIfNeeded() {
        final CollapsingToolbarLayout collapsingToolbarLayout = findViewById(R.id.collapsing_toolbar);
        AppBarLayout appBarLayout = findViewById(R.id.appbar);
        appBarLayout.addOnOffsetChangedListener(new AppBarLayout.OnOffsetChangedListener() {
            boolean isShow = true;
            int scrollRange = -1;

            @Override
            public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
                if (scrollRange == -1) {
                    scrollRange = appBarLayout.getTotalScrollRange();
                }
                if (scrollRange + verticalOffset == 0) {
                    collapsingToolbarLayout.setTitle(currentSeries.getName());

                    isShow = true;
                } else if (isShow) {
                    collapsingToolbarLayout.setTitle(" ");
                    isShow = false;
                }
            }
        });
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;

           /* case R.id.action_delete:
                onDeleteClicked();
                return true;
            case R.id.action_to_watchedlist:
                onAddToWatchedClicked();
                return true;
            case R.id.action_to_watchlist:
                onAddToWatchClicked();
                return true;*/
        }
        return true;
    }
}

File: app/src/main/kotlin/de/cineaste/android/activity/SeasonDetailActivity.kt
package de.cineaste.android.activity

import android.content.Intent
import android.os.Bundle
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.CollapsingToolbarLayout
import androidx.viewpager.widget.ViewPager
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import android.view.MenuItem
import android.widget.ImageView
import com.squareup.picasso.Picasso
import de.cineaste.android.R
import de.cineaste.android.adapter.series.SeasonPagerAdapter
import de.cineaste.android.database.dao.BaseDao
import de.cineaste.android.database.dbHelper.SeriesDbHelper
import de.cineaste.android.entity.series.Series
import de.cineaste.android.util.Constants

class SeasonDetailActivity : AppCompatActivity() {

    private var currentSeries: Series? = null
    private lateinit var poster: ImageView
    private lateinit var seriesDbHelper: SeriesDbHelper

    private var seasonId: Long = 0

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_season_detail)

        seriesDbHelper = SeriesDbHelper.getInstance(this)

        poster = findViewById(R.id.poster_image_view)

        val intent = intent
        val seriesId = intent.getLongExtra(BaseDao.SeasonEntry.COLUMN_SEASON_SERIES_ID, -1)
        seasonId = intent.getLongExtra(BaseDao.SeasonEntry.COLUMN_SEASON_SEASON_NUMBER, -1)

        currentSeries = seriesDbHelper.getSeriesById(seriesId)

        val series = currentSeries

        series?.let { assignData(series) }

        initToolbar()
    }

    private fun assignData(series: Series) {
        val adapter = SeasonPagerAdapter(supportFragmentManager, series, resources)
        val viewPager = findViewById<ViewPager>(R.id.pager)

        viewPager.adapter = adapter
        viewPager.currentItem = currentSeasonIndex()
        viewPager.addOnPageChangeListener(object : ViewPager.OnPageChangeListener {
            override fun onPageScrolled(
                position: Int,
                positionOffset: Float,
                positionOffsetPixels: Int
            ) {
                // do nothing
            }

            override fun onPageSelected(position: Int) {
                setPoster(position)
            }

            override fun onPageScrollStateChanged(state: Int) {
                // do nothing
            }
        })

        setPoster(currentSeasonIndex())
    }

    private fun setPoster(position: Int) {
        currentSeries?.seasons?.let { seasons ->
            val season = seasons[position]

            val posterPath = season.posterPath

            if (posterPath.isNullOrEmpty()) {
                Picasso.get()
                    .load(R.drawable.placeholder_poster)
                    .into(poster)
            } else {
                val posterUri = Constants.POSTER_URI_SMALL
                    .replace("<posterName>", posterPath)
                    .replace("<API_KEY>", getString(R.string.movieKey))
                Picasso.get()
                    .load(posterUri)
                    .error(R.drawable.placeholder_poster)
                    .into(poster)

                poster.setOnClickListener {
                    val intent = Intent(this@SeasonDetailActivity, PosterActivity::class.java)
                    intent.putExtra(PosterActivity.POSTER_PATH, posterPath)
                    startActivity(intent)
                }
            }
        }
    }

    private fun currentSeasonIndex(): Int {
        currentSeries?.seasons?.let { seasons ->
            for (i in seasons.indices) {
                if (seasons[i].id == seasonId) {
                    return i
                }
            }
        }
        return 0
    }

    private fun initToolbar() {
        val toolbar = findViewById<Toolbar>(R.id.toolbar)
        setSupportActionBar(toolbar)
        supportActionBar?.setDisplayHomeAsUpEnabled(true)

        setTitleIfNeeded()
    }

    private fun setTitleIfNeeded() {
        val collapsingToolbarLayout = findViewById<CollapsingToolbarLayout>(R.id.collapsing_toolbar)
        val appBarLayout = findViewById<AppBarLayout>(R.id.appbar)
        appBarLayout.addOnOffsetChangedListener(object : AppBarLayout.OnOffsetChangedListener {
            var isShow = true
            var scrollRange = -1

            override fun onOffsetChanged(appBarLayout: AppBarLayout, verticalOffset: Int) {
                if (scrollRange == -1) {
                    scrollRange = appBarLayout.totalScrollRange
                }
                if (scrollRange + verticalOffset == 0) {
                    val title = currentSeries?.name

                    title?.let {
                        collapsingToolbarLayout.title = title
                    }

                    isShow = true
                } else if (isShow) {
                    collapsingToolbarLayout.title = " "
                    isShow = false
                }
            }
        })
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        when (item.itemId) {
            android.R.id.home -> {
                currentSeries?.let { series ->
                    val unwatchedEpisodes = seriesDbHelper.getUnWatchedEpisodesOfSeries(series.id)
                    if (unwatchedEpisodes.isEmpty() && !series.isInProduction) {
                        series.isWatched = true
                        seriesDbHelper.updateWatchState(series)
                    }
                    onBackPressed()
                    return true
                }
            }
        }

        return super.onOptionsItemSelected(item)
    }
}
